### PR TITLE
Add ability to specify a field within the table's data to use as a row's ID

### PR DESCRIPTION
**Pull request by mollykreis**
# Pull Request

## 🤨 Rationale

Implements the row ID logic described in the table's readme, as modified in #937

## 👩‍💻 Implementation

- Added `id-field-name` attribute
- Added ID validation when the `data` property updates or when the `id-field-name` updates
- Added a distinction between the `data` on the table and the `rows` on the table so that the table could render no rows when necessary
- Added `validity` property and `checkValidity()` function

## 🧪 Testing

- Manually tested in Storybook, including dynamically changing the configuration of the table in dev tools
- Wrote unit tests
    - Note: There are no unit tests that validate that a row's ID is the expected value because the ID is not yet part of the public API of the table. I expect us to add additional ID tests once we have a public API that exposes row IDs. However, this is part of what I manually tested in Storybook.

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: packages/nimble-components/src/table/tests/table.stories.ts [hunk 1]
```diff
@@ -23,10 +23,16 @@ const simpleData = [
...
     [ExampleDataType.simpleData]: simpleData
 } as const;
 
+const dataSetIdFieldNames = {
+    [ExampleDataType.simpleData]: simpleDataIdFieldName
+} as const;
+
 const overviewText = 'The `nimble-table` is a component that offers a way to render tabular data in a variety of ways in each column.';
 
 const dataDescription = `\`data\` is a property that is an array of records. A record provides the data that backs a single row in the table.
```
Comment (jattasNI) [inline packages/nimble-components/src/table/tests/table.stories.ts:38]: Thoughts on documenting `id-field-name` and the new validity APIs in storybook? The client can't set all of them so they might not make sense as controls, but I can't think of a better place to make clients aware of them.

In a perfect world our docs would provide both normal API info (data types, purpose of the API) and also info about why someone might choose to set a custom ID field and when we'd recommend they perform validity checks. If we don't know the answers to all of those yet we can start with what we do know.

File: packages/nimble-components/src/table/index.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ import {
 } from '@tanstack/table-core';
 import { styles } from './styles';
 import { template } from './template';
-import type { TableRecord } from './types';
+import type { TableRecord, TableRowState, TableValidity } from './types';
```
Comment (jattasNI) [inline packages/nimble-components/src/table/index.ts:14]: Do we need to export these types from `nimble-angular` [like we do for other types](https://github.com/ni/nimble/blob/86302af34ab8c23201b56001fe9efbd599e46878/angular-workspace/projects/ni/nimble-angular/src/directives/table/nimble-table.directive.ts#L6)?

File: packages/nimble-components/src/table/index.ts [hunk 3]
```diff
@@ -66,15 +103,71 @@ export class Table<
...
+            return true;
+        }
+
+        const ids = new Set<string>();
+        for (const record of this.data) {
+            if (
+                !Object.prototype.hasOwnProperty.call(record, this.idFieldName)
+            ) {
+                this.missingRowId = true;
+                return false;
```
Comment (jattasNI) [inline packages/nimble-components/src/table/index.ts:139]: Is there value in not returning early in case the ids have multiple problems?

File: packages/nimble-components/src/table/index.ts [hunk 4]
```diff
@@ -66,15 +103,71 @@ export class Table<
...
+            }
+
+            const id = record[this.idFieldName];
```
Comment (jattasNI) [inline packages/nimble-components/src/table/index.ts:142]: If you haven't already, could you double check corner cases like `id === null` and `id === ''`? I'm guessing `null` would result in `invalidRowId` and `''` would be a valid id as long as no other record also uses it, but JS type rules and Set behavior might produce a different outcome.

**Review by jattasNI — APPROVED** [review]

File: packages/nimble-components/src/table/index.ts [hunk 5]
```diff
@@ -37,16 +46,35 @@ export class Table<
...
+        };
+    }
+
     private readonly table: TanStackTable<TData>;
     private options: TanStackTableOptionsResolved<TData>;
     private readonly tableInitialized: boolean = false;
 
+    private duplicateRowId = false;
+    private missingRowId = false;
+    private invalidRowId = false;
```
Comment (rajsite) [inline packages/nimble-components/src/table/index.ts:63]: thought on breaking this off into a separate class like `TableValidator` in a spot like `table/models/table-validator` to capture the different properties and generate the TableValidity object? Seems like it could be a pretty self contained class with methods like reset(), validateData(), getTableValidity(), and isValid(), etc. Could also split out its unit tests.

Comment (rajsite) [reply packages/nimble-components/src/table/index.ts:63]: Compared to other controls I'm more concerned about the table class getting really big

File: packages/nimble-components/src/table/index.ts (hunk 3 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:139]: The value is that we aren't doing extra work when we already know that the configuration is invalid, but I don't mind updating this to report all the problems. I was on the fence myself about this.

Comment (jattasNI) [reply packages/nimble-components/src/table/index.ts:139]: I also don't feel too strongly, but I think I'd vote to continue validation beyond the first error. In the case we're considering, the table's already in a bad state so performance is not too important. And if we can save a developer from multiple rounds of debugging to discover all the errors then that's worthwhile.

File: packages/nimble-components/src/table/index.ts (hunk 4 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:142]: I think we'll actually run into problems with a client trying to have one of their IDs be an empty string (or null). If our `getRowId` function returns a falsey value, TanStack will generate a value for the ID. I'll add a check in the code for this along with a test.

File: packages/nimble-components/src/table/index.ts (hunk 2 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:14]: Yes, we will need to export `TableValidity` from Angular. I'll do that in a separate PR because I also need to add support for the `id-field-name` attribute, `validity` property, and `checkValidity()` function.

`TableRowState` is intended to be internal within the table even though `tableData` is required to be public in order for the template to have access to it. I don't think we need to expose that type through to Angular.

File: packages/nimble-components/src/table/tests/table.stories.ts (hunk 1 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/tests/table.stories.ts:43]: I've documented the behavior `id-field-name`, `validity`, and `checkValidity()`, but I didn't add any usage recommendations. I think we'll need to work on defining what our recommendations are.

File: packages/nimble-components/src/table/index.ts (hunk 3 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts]: Updated & added a test

File: packages/nimble-components/src/table/index.ts (hunk 5 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts]: I've added a `TableValidator` class with a bunch of tests. I removed some of the tests from the table component itself, but I kept a few so that we could validate that the validity state bubbles up correctly through the component and that the component renders as expected based on whether or not the ID is valid.

File: packages/nimble-components/src/table/index.ts (hunk 4 above)
Comment (rajsite) [reply packages/nimble-components/src/table/index.ts]: I had a related [comment](https://github.com/ni/nimble/pull/947#discussion_r1066447569) in @atmgrifter00's PR (i.e. Is empty string actually an invalid record name?).

Can we make sure to clarify the edge cases around empty strings as field names / id values and maybe capture it in specs and tests?

- are record field names allowed to be empty string? `table.data= [{"": 1}, {"": 2}]`
- is the idFieldName allowed to be empty string? `table.data= [{"": 1}, {"": 2}]; table.idFieldName = ''`
- is empty string a valid id for id value type string? `table.data= [{id: "a"}, {id: ""}]; table.idFieldName = 'id'`
- are ieee754 values allowed for id for type number? `table.data= [{id: 1}, {id: Infinity}, {id: -Infinity}, , {id: -0}, {id: +0}, {id: NaN}]; table.idFieldName = 'id'`
  - I think -Inf and +Inf are fine as values
  - We should not distinguish between -0 and +0, they should be treated as the same value
    - ie. `Object.is(-0, 0) // false` while `-0 === 0 // true`
  - NaN should not be allowed as a value for id
  - We should not allow id values outside the safe integer range
  - Edit: Actually maybe we should just only allow integer values as number ids? And enforce they are all safe with [isSafeInteger()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger#:~:text=Otherwise%20false%20.-,Description,%2C%20inclusive%20(%C2%B19%2C007%2C199%2C254%2C740%2C991).)
    - I can maybe see this being important if the integers are coming from the server as 64-bit int values by accident which cannot be safely represented as JS numbers.
- what types are allowed for the id?
    - I assume definitely: string and number
    - I assume not: null, undefined, boolean
    - Should date? Seems to have the characteristics to allow for it.

Comment (rajsite) [reply packages/nimble-components/src/table/index.ts]: As this is mostly edge cases and tests, I'll assume we won't address it in this PR but handle in a follow-up

**Review by atmgrifter00 — APPROVED** [review]

File: packages/nimble-components/src/table/index.ts [hunk 6]
```diff
@@ -66,15 +97,39 @@ export class Table<
 
         // Ignore any updates that occur prior to the TanStack table being initialized.
         if (this.tableInitialized) {
-            this.updateTableOptions({ data: this.data });
+            this.trySetData(this.data);
```
Comment (rajsite) [inline packages/nimble-components/src/table/index.ts:100]: Say for example the user has an array assigned to data, they mutate the data and assign the same array again (to trigger an update). Would we need the shallow copy like the idFieldNameChange above for tanstack to detect?

Maybe after validation, trySetData should always create a shallow copy to pass to updateTableOptions?

```ts
const arr = []; // good data
table.data = arr
table.idFieldName = '' // good id
// sometime later
arr[0] //mutate but still good
table.data = arr // do assignment to trigger update
```

File: packages/nimble-components/src/table/types.ts [hunk 7]
```diff
@@ -24,3 +24,13 @@ export type TableFieldValue =
...
 }
+
+export interface TableValidity {
+    readonly duplicateRowId: boolean;
+    readonly missingRowId: boolean;
+    readonly invalidRowId: boolean;
+}
+
+export interface TableRowState<TData extends TableRecord = TableRecord> {
+    data: TData;
```
[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/types.ts:35]: I had `id` as part of this interface, but then I decided to remove it since it wasn't needed yet. Ultimately `id` will be in this interface though. I can add it back now if you think that would be better.

File: packages/nimble-components/src/table/index.ts (hunk 4 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts]: To answer some of the questions you've posed here:
- This PR assumes that record field names can be an empty string. Based on a previous comment in this PR, I thought that was what we wanted. I have a test to ensure the empty field name works correctly from a validation point-of-view. I manually tested that it works on the table, but there isn't an auto test for it.
- TanStack does not accept an empty string for the ID, so I don't see a good path forward of accepting this as a valid ID on the `nimble-table`. I have validation code & a corresponding test for this. I also included this in the docs.
- TanStack only allows strings for the row IDs. I'd prefer not to do conversion between the `nimble-table` IDs and TanStack IDs, so I think we should stick with requiring string IDs unless there is a strong reason that we need support another data type.

File: packages/nimble-components/src/table/index.ts [hunk 8]
```diff
@@ -66,15 +97,39 @@ export class Table<
...
+        } else {
+            this.updateTableOptions({ data: [] });
+        }
+    }
+
+    private refreshRows(): void {
+        const rows = this.table.getRowModel().rows;
+        this.tableData = rows.map(row => {
+            return { data: row.original, id: row.id } as TableRowState<TData>;
+        });
```
Comment (rajsite) [inline packages/nimble-components/src/table/index.ts:124]: I'm assuming this type is actually compatible with TableRowState, instead of the `as` cast (which disables checks and just asserts) can we leverage the type?

```suggestion
        this.tableData = rows.map(row => {
            const rowState: TableRowState<TData> = { data: row.original, id: row.id };
            return rowState;
        });
```

File: packages/nimble-components/src/table/models/table-validator.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        this.missingRowId = false;
+        this.invalidRowId = false;
+
+        if (idFieldName == null) {
+            return true;
+        }
+
+        const ids = new Set<string>();
+        for (const record of data) {
+            if (!Object.prototype.hasOwnProperty.call(record, idFieldName)) {
```
Comment (rajsite) [inline packages/nimble-components/src/table/models/table-validator.ts:39]: nit: [Object.hasOwn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn) is a shortcut for exactly this

**Review by rajsite — APPROVED** [review]

File: packages/nimble-components/src/table/index.ts (hunk 6 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:100]: Currently, `data` is an `@observable` property on the table and the code to update TanStack happens within `dataChanged`. I'm not aware of a scenario where FAST would detect a change to the array but TanStack would not.

Perhaps it is better to create a setter for `data` rather than have it be observable, particularly because the template no longer binds to `data` (now it uses `tableData`). I'll play around with this, and assuming a setter gives us better behavior, I'll post a follow-up PR to change the approach.

File: packages/nimble-components/src/table/index.ts (hunk 8 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:124]: Your suggestion actually caught a legitimate issue with this line, so I'll fix that too.

File: packages/nimble-components/src/table/index.ts (hunk 6 above)
Comment (rajsite) [reply packages/nimble-components/src/table/index.ts:100]: > I'm not aware of a scenario where FAST would detect a change to the array but TanStack would not.

I think then I don't fully understand why the shallow copy is needed for calling trySetData in idFieldNameChanged but not for calling trySetData in dataChanged.

My mental model is that calling updateTableOptions internally was just checking if the data reference was pointing to the same object or something and creating a shallow copy kicked it into recognizing a change by pointing to a new one. If that's the case then I would imagine we need a shallow copy in both cases.

Just trying to make sure it's clear when we need a shallow copy and maybe the comment could clarify a bit when it's needed.

File: packages/nimble-components/src/table/models/table-validator.ts (hunk 9 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/models/table-validator.ts:39]: I get this error when I try that:
`TS2550: Property 'hasOwn' does not exist on type 'ObjectConstructor'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2022' or later.`

File: packages/nimble-components/src/table/types.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply packages/nimble-components/src/table/types.ts:35]: Your call. I just noticed that we were already returning `TableRowState` instances with the `id` set, so it seemed reasonable for it to be on the interface.

[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/types.ts:35]: I had just noticed that too. For right or for wrong, I went ahead and removed setting `id` when creating the `TableRowState`.

File: packages/nimble-components/src/table/models/table-validator.ts (hunk 9 above)
Comment (rajsite) [reply packages/nimble-components/src/table/models/table-validator.ts:39]: eh no biggie for now, it's just syntax sugar, can make a concerted effort to pickup some new es2022 syntax later

File: packages/nimble-components/src/table/index.ts (hunk 6 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:100]: The reason that a shallow copy isn't needed in the code path where the data changes is because the current implementation requires the client to do that shallow copy. Otherwise, FAST won't detect any changes to the array, and our code to update TanStack won't get hit at all. With the current table code in master (unchanged with this PR), the last line in your code snippet above won't actually trigger an update. The client would need to write `table.data = [ ...arr ]` and do the shallow copy themselves.

My proposal is to (in a separate PR) make the `data` property on the table have a getter/setter rather than rely on FAST's change detection on the array. With that change, we will need to do the shallow copy both when the data is set and when the id changes, and the client will no longer be required to do so.

Comment (rajsite) [reply packages/nimble-components/src/table/index.ts:100]: perfect, thanks for the clarification! 🚢-it!

Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:100]: I created #958 so we wouldn't lose track of needing to make this change.
